Source/WebCore/platform/mediastream/MediaStreamTrackPrivate.cpp

```diff
@@ -173,44 +443,57 @@ Ref<MediaStreamTrackPrivate> MediaStreamTrackPrivate::clone()
     clonedMediaStreamTrackPrivate->m_contentHint = this->m_contentHint;
     clonedMediaStreamTrackPrivate->updateReadyState();
 
-    if (m_source->isProducingData())
+    if (m_isProducingData)
         clonedMediaStreamTrackPrivate->startProducingData();
 
     return clonedMediaStreamTrackPrivate;
 }
 
-const RealtimeMediaSourceSettings& MediaStreamTrackPrivate::settings() const
+RealtimeMediaSource& MediaStreamTrackPrivate::source()
 {
-    return m_source->settings();
+    ASSERT(isMainThread());
+    return m_sourceObserver->source();
 }
```

Comment:
It would be great to remove direct access to the source completely

Reply:
Yes, I plan to do some further refactoring, we will see how far we can go.